File: AIGAProject/Model/Maps.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace AIGAProject.Model
{
    class Map
    {
        //老師給定的，寫死
        const int EMPTY_CELL = 0;
        const int OBSTACLE_CELL = 1;
        const int START_CELL = 2;
        const int GOAL_CELL = 3;
        int[,] mapArray;

        public int Height
        {
            get;
        }

        public int Width
        {
            get;
        }
        public Point StartPoint
        {
            get;
        }

        public Point GoalPoint
        {
            get;
        }

        bool IsObstacle(Point point)
        {
            return mapArray[point.Y, point.X] == OBSTACLE_CELL;
        }

        Point MoveNorthWay(Point startLocation, Step step)
        {
            Point point = startLocation;
            for (int i = 0; i < step.Count; i++)
            {
                Point nextPoint = point;
                nextPoint.Y++;
                if (OutOfBound(nextPoint))
                {
                    return point;
                }
                if (!IsObstacle(nextPoint))
                {
                    point = nextPoint;
                }
            }
            return point;
        }

        Point MoveNorthEastWay(Point startLocation, Step step)
        {
            Point point = startLocation;
            for (int i = 0; i < step.Count; i++)
            {
                Point nextPoint = point;
                nextPoint.X++;
                nextPoint.Y++;
                if (OutOfBound(nextPoint))
                {
                    return point;
                }
                if (!IsObstacle(nextPoint))
                {
                    point = nextPoint;
                }
            }
            return point;
        }

        Point MoveEastWay(Point startLocation, Step step)
        {
            Point point = startLocation;
            for (int i = 0; i < step.Count; i++)
            {
                Point nextPoint = point;
                nextPoint.X++;
                if (OutOfBound(nextPoint))
                {
                    return point;
                }
                if (!IsObstacle(nextPoint))
                {
                    point = nextPoint;
                }
            }
            return point;
        }

        Point MoveSouthEastWay(Point startLocation, Step step)
        {
            Point point = startLocation;
            for (int i = 0; i < step.Count; i++)
            {
                Point nextPoint = point;
                nextPoint.X++;
                nextPoint.Y--;
                if (OutOfBound(nextPoint))
                {
                    return point;
                }
                if (!IsObstacle(nextPoint))
                {
                    point = nextPoint;
                }
            }
            return point;
        }

        Point MoveSouthWay(Point startLocation, Step step)
        {
            Point point = startLocation;
            for (int i = 0; i < step.Count; i++)
            {
                Point nextPoint = point;
                nextPoint.Y--;
                if (OutOfBound(nextPoint))
                {
                    return point;
                }
                if (!IsObstacle(nextPoint))
                {
                    point = nextPoint;
                }
            }
            return point;
        }

        Point MoveSouthWestWay(Point startLocation, Step step)
        {
            Point point = startLocation;
            for (int i = 0; i < step.Count; i++)
            {
                Point nextPoint = point;
                nextPoint.X--;
                nextPoint.Y--;
                if (OutOfBound(nextPoint))
                {
                    return point;
                }
                if (!IsObstacle(nextPoint))
                {
                    point = nextPoint;
                }
            }
            return point;
        }

        Point MoveWestWay(Point startLocation, Step step)
        {
            Point point = startLocation;
            for (int i = 0; i < step.Count; i++)
            {
                Point nextPoint = point;
                nextPoint.X--;
                if (OutOfBound(nextPoint))
                {
                    return point;
                }
                if (!IsObstacle(nextPoint))
                {
                    point = nextPoint;
                }
            }
            return point;
        }

        Point MoveNorthWestWay(Point startLocation, Step step)
        {
            Point point = startLocation;
            for (int i = 0; i < step.Count; i++)
            {
                Point nextPoint = point;
                nextPoint.X--;
                nextPoint.Y++;
                if (OutOfBound(nextPoint))
                {
                    return point;
                }
                if (!IsObstacle(nextPoint))
                {
                    point = nextPoint;
                }
            }
            return point;
        }

        bool OutOfBound(Point p)
        {
            return (p.X < 0 ||
                    p.Y < 0 ||
                    p.X >= Width ||
                    p.Y >= Height);
        }

        public Point GetLocationResult(Point startLocation, Step step)
        {
            switch (step.Direction)
            {
                case Direction.N:
                    return MoveNorthWay(startLocation, step);                    
                case Direction.NE:
                    return MoveNorthEastWay(startLocation, step);                    
                case Direction.E:
                    return MoveEastWay(startLocation, step);                    
                case Direction.SE:
                    return MoveSouthEastWay(startLocation, step);                    
                case Direction.S:
                    return MoveSouthWay(startLocation, step);                    
                case Direction.SW:
                    return MoveSouthWestWay(startLocation, step);                    
                case Direction.W:
                    return MoveWestWay(startLocation, step);                    
                case Direction.NW:
                    return MoveNorthWestWay(startLocation, step);                    
            }
            return startLocation;
        }

        public Map(string path)
        {
            using (var reader = new StreamReader(path))
            {
                int j = 0;
                while (!reader.EndOfStream)
                {
                    var line = reader.ReadLine();
                    var values = line.Split(',');

                    for (int i = 0; i < values.Length; i++)
                    {
                        Height = values.Length;
                        Width = values.Length;
                        mapArray = new int[values.Length, values.Length];
                        mapArray[j, i] = int.Parse(values[i]);
                        switch (mapArray[j, i])
                        {
                            case START_CELL:
                                StartPoint = new Point(j, i);
                                break;
                            case GOAL_CELL:
                                GoalPoint = new Point(j, i);
                                break;
                        }
                    }
                    j++;
                }
            }
        }
    }

    class Maps
    {
        Map testMap = new Map("../../../../ai_map/mytest.csv");

        public Map TestMap
        {
            get
            {
                return testMap;
            }
        }
    }
}

File: AIGAProject_py/Maps.py
from typing import List
from Struct.Point import Point
from MyException import MyException
from enum.MapType import MapType

class Map:
    EMPTY_CELL: int = 0
    OBSTACLE_CELL: int = 1
    START_CELL: int = 2
    GOAL_CELL: int = 3

    def __init__(self, path: str) -> None:
        _height: int = 0
        _width: int = 0
        _start_point: Point = None
        _goal_point: Point = None
        map_array: List[int] = None
        
        with open(path, "r", encoding="utf-8") as file:
            for j, line in file:
                line = file.readline()
                values = line.split(",")

                height: int = len(values)
                width: int = len(values)
                for i in range(len(values)):
                    map_array = [[] * width] * height
                    int_value = int(value)
                    map_array[j][i] = int_value
                    if int_value == self.START_CELL:
                        self._start_point = Point(i, j)
                    elif int_value == self.GOAL_CELL:
                        self._goal_point = Point(i, j)
                

    @property
    def Height(self) -> int:
        return self._height
        
    @property
    def Width(self) -> int:
        return self._width
        
    @property
    def StartPoint(self) -> Point:
        return self._start_point
        
    @property
    def GoalPoint(self) -> Point:
        return self._goal_point

    def out_of_bound(self, p: Point, robot_radius: int) -> bool:
        return (
            p.X - robot_radius < 0 
            or p.Y - robot_radius < 0 
            or p.X + robot_radius >= self.Width 
            or p.Y + robot_radius >= self.Height
        )

    def is_obstacle(self, point: Point, robot_radius: int):
        return map_array[point.Y + robot_radius, point.X] == OBSTACLE_CELL or
            map_array[point.Y + robot_radius, point.X + robot_radius]  == OBSTACLE_CELL or
            map_array[point.Y, point.X + robot_radius]  == OBSTACLE_CELL or
            map_array[point.Y - robot_radius, point.X + robot_radius]  == OBSTACLE_CELL or
            map_array[point.Y - robot_radius, point.X] == OBSTACLE_CELL or
            map_array[point.Y - robot_radius, point.X - robot_radius]  == OBSTACLE_CELL or
            map_array[point.Y, point.X - robot_radius]  == OBSTACLE_CELL or
            map_array[point.Y + robot_radius, point.X - robot_radius]  == OBSTACLE_CELL

    def location_vaild(self, next_point: Point, robot_radius: int) -> bool:
        out_of_bound_check = out_of_bound(nextPoint, robot_radius)
        is_obstacle_check = is_obstacle(nextPoint, robot_radius)
        return not out_of_bound_check and not is_obstacle_check


class Maps:
    def __init__(self) -> None:
        self.test_map = Map("../../../../ai_map/mytest.csv")

    def get_map(self, map_type: MapType) -> Map:
        if map_type == MapType.Test:
            return self.test_map
        raise MyException()

